infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/OnWorkspaceRemoveDataVolumeRemover.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.docker;
+
+import static java.lang.String.format;
+import static org.slf4j.LoggerFactory.getLogger;
+
+import java.io.IOException;
+import java.util.List;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.core.notification.EventSubscriber;
+import org.eclipse.che.api.workspace.shared.event.WorkspaceRemovedEvent;
+import org.eclipse.che.infrastructure.docker.client.DockerConnector;
+import org.eclipse.che.infrastructure.docker.client.json.volume.Volume;
+import org.eclipse.che.workspace.infrastructure.docker.provisioner.volume.VolumeNames;
+import org.slf4j.Logger;
+
+/** @author Alexander Garagatyi */
+@Singleton
+public class OnWorkspaceRemoveDataVolumeRemover implements EventSubscriber<WorkspaceRemovedEvent> {
+
+  private static final Logger LOG = getLogger(OnWorkspaceRemoveDataVolumeRemover.class);
+
+  private final DockerConnector docker;
+
+  @Inject
+  public OnWorkspaceRemoveDataVolumeRemover(DockerConnector docker) {
+    this.docker = docker;
+  }
+
+  @Override
+  public void onEvent(WorkspaceRemovedEvent event) {
+    String wsId = event.getWorkspace().getId();
+    try {
+      List<Volume> volumes = docker.getVolumes().getVolumes();
```

Comment:
call `docker.getVolumes().getVolumes()` looks strange for me 
consider to change it to something like: `docker.getVolumes().asList()` / `docker.getVolumes().values()` / `docker.getVolumes().items()`

Reply:
I don't like it either. But in this case, I would have to customize POJO `Volumes` with annotations to customize parsing logic. And I don't want to add this complexity, so for me, the current code is an acceptable tradeoff.

Reply:
in docker API:
get volumes returns
```
{
    "volumes": []
}
```

Reply:
I see, let it be as it is